packages/x-date-pickers-pro/src/DateTimeRangePicker/shared.tsx

```diff
@@ -0,0 +1,194 @@
+import {
+  MultiSectionDigitalClockSlots,
+  MultiSectionDigitalClockSlotProps,
+} from '@mui/x-date-pickers/MultiSectionDigitalClock/MultiSectionDigitalClock.types';
+import { resolveTimeViewsResponse } from '@mui/x-date-pickers/internals/utils/date-time-utils';
+import { UseViewsOptions } from '@mui/x-date-pickers/internals/hooks/useViews';
+import { DateTimeRangeValidationError } from '../models';
+import {
+  DateRange,
+  DateTimeRangePickerView,
+  DateTimeRangePickerViewExternal,
+} from '../internals/models';
+import {
+  DateRangeCalendarSlots,
+  DateRangeCalendarSlotProps,
+  ExportedDateRangeCalendarProps,
+} from '../DateRangeCalendar';
+import {
+  DateTimeRangePickerToolbar,
+  DateTimeRangePickerToolbarProps,
+  ExportedDateTimeRangePickerToolbarProps,
+} from './DateTimeRangePickerToolbar';
+import { DateRangeViewRendererProps } from '../dateRangeViewRenderers';
+import {
+  DateTimeRangePickerTabs,
+  DateTimeRangePickerTabsProps,
+  ExportedDateTimeRangePickerTabsProps,
+} from './DateTimeRangePickerTabs';
+
+export interface BaseDateTimeRangePickerSlots<TDate>
+  extends DateRangeCalendarSlots<TDate>,
+    DigitalClockSlots,
+    MultiSectionDigitalClockSlots {
+  /**
+   * Tabs enabling toggling between date and time pickers.
+   * @default DateTimeRangePickerTabs
+   */
+  tabs?: React.ElementType<DateTimeRangePickerTabsProps>;
+  /**
+   * Custom component for the toolbar rendered above the views.
+   * @default DateTimeRangePickerToolbar
+   */
+  toolbar?: React.JSXElementConstructor<DateTimeRangePickerToolbarProps<TDate>>;
+}
+
+export interface BaseDateTimeRangePickerSlotProps<TDate>
+  extends DateRangeCalendarSlotProps<TDate>,
+    DigitalClockSlotProps,
+    MultiSectionDigitalClockSlotProps {
+  /**
+   * Props passed down to the tabs component.
+   */
+  tabs?: ExportedDateTimeRangePickerTabsProps;
+  /**
+   * Props passed down to the toolbar component.
+   */
+  toolbar?: ExportedDateTimeRangePickerToolbarProps;
+}
+
+export interface BaseDateTimeRangePickerProps<TDate>
+  extends Omit<
+      BasePickerInputProps<
+        DateRange<TDate>,
+        TDate,
+        DateTimeRangePickerView,
+        DateTimeRangeValidationError
+      >,
+      'orientation' | 'views' | 'openTo'
+    >,
+    Omit<ExportedDateRangeCalendarProps<TDate>, 'calendars'>,
+    BaseDateValidationProps<TDate>,
+    DesktopOnlyTimePickerProps<TDate>,
+    Partial<
+      Pick<UseViewsOptions<DateRange<TDate>, DateTimeRangePickerViewExternal>, 'openTo' | 'views'>
+    > {
+  /**
+   * Overridable component slots.
+   * @default {}
+   */
+  slots?: BaseDateTimeRangePickerSlots<TDate>;
+  /**
+   * The props used for each component slot.
+   * @default {}
+   */
+  slotProps?: BaseDateTimeRangePickerSlotProps<TDate>;
+  /**
+   * Define custom view renderers for each section.
+   * If `null`, the section will only have field editing.
+   * If `undefined`, internally defined view will be the used.
+   */
+  viewRenderers?: Partial<
```

Comment:
This type is probably worth extracting in a standalone type

Reply:
It does not seem to be re-used anywhere else currently. 🤔 
What value do you see in extracting? 🤔

Reply:
Readability mostly

Reply:
I would argue that having the type inlined while it's not used elsewhere increases the readability because there is no need to jump between code lines or even files. 🤔 
But we do often extract more complex types so, we can do it here as well. 👍

Reply:
Feel free to ignore this proposal if you think it does not increase readability :+1: